### DISCUSSION TITLE
After build for linux, Dist folder contains py file in plain text - Is it correct ?

### DISCUSSION THREAD

**Discussion by gikpro** [pyinstaller/pyinstaller#6926] [Help]
I am trying to make my program works on Linux system (Xubuntu 20.04.4 LTS (Focal Fossa), I use Kivy and kivymd, i have many modules (components, utils, screens, kv) that contains .py files and .kv files, i have also a images folder, that are not detected automaticaly by pyinstaller so i have added them in "datas" parameters of Analysis(), The program doesn't show import error after the build, but now i see that my python file in kv file in plain text(everyone can see my code). How i can resolve that.

Here my spec
```
# -*- mode: python ; coding: utf-8 -*-
import sys
import os

from kivymd import hooks_path as kivymd_hooks_path 

path = os.path.abspath(".")

screens_module = ("screens/","screens")
kv_module = ("kv/","kv")
utils_module = ("utils/","utils")
components_module = ("components/","components")
images_module = ("images/","images")
added_files = [screens_module, kv_module, utils_module, components_module, images_module]
block_cipher = None

a = Analysis(
    ['main.py'],
    pathex=[path],
    datas=added_files,
    hiddenimports=[],
    hookspath=[kivymd_hooks_path],
    hooksconfig={},
    runtime_hooks=[], 
    excludes=['tkinter'],
    win_no_prefer_redirects=False,
    win_private_assemblies=False,
    cipher=block_cipher,
    noarchive=False,
    
)
pyz = PYZ(a.pure, a.zipped_data, cipher=block_cipher)

exe = EXE(
    pyz,
    a.scripts,
    [],
    exclude_binaries=True,
    name='haitinrv-linux',
    debug=False,
    bootloader_ignore_signals=False,
    strip=False,
    upx=True,
    console=True,
    disable_windowed_traceback=False,
    argv_emulation=False,
    target_arch=None,
    codesign_identity=None,
    entitlements_file=None,
)
coll = COLLECT(
    exe,
    a.binaries,
    a.zipfiles,
    a.datas,
    strip=False,
    upx=True,
    upx_exclude=[],
    name='haitinrv-linux',
)

```

## Dist folder generated

![dist-folder-linux](https://user-images.githubusercontent.com/77642887/177013566-164241e7-82c1-4f91-bd97-fd13e025d3d2.png)

### Py files in dist folder in the home news module
![dist-screens-homenews-pyfile](https://user-images.githubusercontent.com/77642887/177013641-29f8b005-be44-4560-8b07-6ecd56c25182.png)

Here  the hierarchy of my code

![haiti-nrv-root](https://user-images.githubusercontent.com/77642887/177013364-be2c3a1f-5f80-4ce0-846f-e32344d0ffae.png)

## Screens module

![screens](https://user-images.githubusercontent.com/77642887/177013436-e182dfe3-7d75-4f01-9370-cb657486b5c8.png)

### Screens module - home news module

![screens-home_news](https://user-images.githubusercontent.com/77642887/177013463-81b3ec2f-3c98-48e2-9502-bbf18e4c977c.png)
 
## Kv module

![kv](https://user-images.githubusercontent.com/77642887/177013494-43759a2b-52c7-4883-a76f-77c15c56f6cb.png)

### Kv module - Home news

![kv-home_news](https://user-images.githubusercontent.com/77642887/177013506-fc5c577d-0515-4496-9141-f28dd89a19d4.png)

## Components
![components-kv](https://user-images.githubusercontent.com/77642887/177013625-483d905c-5ea6-480b-8092-d0e5e3e75a4b.png)

**Comment by bwoodsend** [comment]
If `components`, `utils`, `screens` and `kv` directories contain `.py` files and you tell PyInstaller to add those directories as data files then of course you're going to have `.py` files in the output.

**Comment by rokm** [comment]
> ```python
> screens_module = ("screens/","screens")
> kv_module = ("kv/","kv")
> utils_module = ("utils/","utils")
> components_module = ("components/","components")
> images_module = ("images/","images")
> added_files = [screens_module, kv_module, utils_module, components_module, images_module]
> # ...
> 
> a = Analysis(
>     ['main.py'],
>     pathex=[path],
>     datas=added_files,
>     hiddenimports=[],
>     # ...
> ```

You are explicitly asking PyInstaller to take those directories (`screens`, `kv`, `utils`, `components`) and treat them as plain data - so they are copied over, along with all their contents. 

To collect .kv files, you should programmatically list the files and add the entries to `datas` (or perhaps use the [`Tree` class](https://pyinstaller.org/en/stable/advanced-topics.html#the-tree-class) to list contents of the directories). If this was an installed package, you could use [`collect_data_files` hook utility function](https://pyinstaller.org/en/stable/hooks.html#PyInstaller.utils.hooks.collect_data_files), but when these top-level packages are located next to entry-point script, you would have to ensure that the project directory is in `sys.path`, otherwise the packages won't be found when `collect_data_files` is called from .spec file before `Analysis` is instantiated.

If you need to also explicitly collect python modules from those directories, you should add module names (not filenames!) to `hiddenimports`. But this should be necessary only if there are indirect, lazy, or programmatic imports. Again, if this was an installed package, you could use [`collect_submodules()`](https://pyinstaller.org/en/stable/hooks.html#PyInstaller.utils.hooks.collect_submodules) to walk each top-level package and collect all its submodules. In this case, you would need to ensure that top-level directory is in `sys.path` before calling the hook utility functions.
